cachito/workers/pkg_managers/general_js.py

```diff
@@ -44,6 +51,67 @@
 log = logging.getLogger(__name__)
 
 
+def parse_dependency(proxy_repo_url, dep_identifier):
+    """
+    Parse the dependecy to return the proxied download URL and the file name.
+
+    :param str proxy_repo_url: The Nexus proxy repository URL to use as the registry
+    :param str dep_identifier:
+    :return: proxied_url for the download and the tarball_name.
+    :rtype: str, str
+    """
+    full_pkg_name, version = dep_identifier.rsplit("@", 1)
+    if "/" in full_pkg_name:
+        short_pkg_name = full_pkg_name.rsplit("/", 1)[1]
+    else:
+        short_pkg_name = full_pkg_name
+
+    proxied_url = f"{proxy_repo_url}{full_pkg_name}/-/{short_pkg_name}-{version}.tgz"
+    tarball_name = f"{full_pkg_name}-{version}.tgz".replace("@", "").replace("/", "-")
+
+    return proxied_url, tarball_name
+
+
+async def get_dependencies(proxy_repo_url, download_dir, deps_to_download):
+    """
+    Asynchronous function that execute the dependencies download.
+
+    Receives the url (proxy_repo_url), the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param str proxy_repo_url: The Nexus proxy repository URL to use as the registry.
+    :param str download_dir: Path to download file to.
+    :param list[str] deps_to_download: List of dependencies to be downloaded.
+    :return: a list of the downloaded tarballs.
+    :rtype: list[str]
+    """
+    nexus_username, nexus_password = nexus.get_nexus_hoster_credentials()
+    nexus_auth = aiohttp.BasicAuth(nexus_username, nexus_password)
+
+    concurrency_limit = asyncio.Semaphore(5)
+
+    async with aiohttp.ClientSession() as session:
+
+        tasks = []
+        for dep_identifier in deps_to_download:
+            proxied_url, tarball_name = parse_dependency(proxy_repo_url, dep_identifier)
+            tasks.append(
+                asyncio.ensure_future(
+                    async_download_binary_file(
+                        session,
+                        proxied_url,
```

Comment:
Very late realization on my part, but I'm not sure if Nexus will have the metadata `npm install` needs if we just download the tarball directly.

I think that for python, Nexus caches both the HTTP response from `/simple/$package_name/` and the tarball itself. I assume something similar happens with npm.

Was this ever verified during the design? I'm thinking how to check it. Maybe just try to install from the outputs of a request

Reply:
I don't think this factor was considered at any point in the design. How can I do this test?

Reply:
Some documentation is in https://github.com/containerbuildsystem/cachito/blob/master/docs/using_requests_locally.md, but when I tried with an integration test scenario, the local install didn't work even with the master branch...

The most reliable way to test it would probably be OSBS integration tests - after deploying to stage :sweat_smile:

Reply:
So I dove a bit deeper into this and I don't think it should be an issue.

After hunting down the relevant part of the npm CLI code, I found something called a "packument": https://github.com/npm/pacote/blob/main/lib/registry.js#L38. I believe that's the metadata npm needs.

For example, for `chai` (one of the deps in our integration tests), the packument would be this: https://registry.npmjs.org/chai

So I cleaned up my local env, ran your updated version, submitted a Cachito request and checked if the packument was there:
```shell
$ grep -R --files-with-matches "BDD/TDD assertion library for node.js and the browser. Test framework agnostic." tmp/nexus-data/blobs
tmp/nexus-data/blobs/cachito-js/content/vol-22/chap-03/c61d5874-d4a9-4460-9216-a0a2f375d7b4.bytes
tmp/nexus-data/blobs/cachito-js/content/vol-15/chap-03/f89d530d-ff9f-4f57-ae9e-f2753d3a2243.bytes
tmp/nexus-data/blobs/cachito-js/content/vol-33/chap-16/c1490d97-6d61-4320-8eb8-8f519a507a79.bytes

$ jq < tmp/nexus-data/blobs/cachito-js/content/vol-33/chap-16/c1490d97-6d61-4320-8eb8-8f519a507a79.bytes
# indeed that's the packument!
```
...in fact, the only notable difference between the master branch and this branch (as far as the blobs are concerned) is that the master branch downloads 3 copies of the [npm packument](https://registry.npmjs.org/npm) for no reason (and it's massive, something like 19 Mi each). So I think we have all the metadata we need and even get rid of unwanted extras

Reply:
> 3 copies of the [npm packument](https://registry.npmjs.org/npm) for no reason

You know what it could be? The `npm pack` call probably checks for updates and has to download the npm packument to do that. Let's assume missing it is OK, it should be fairly easy to fix if we do need it

Reply:
> > @MartinBasti
> > ![image](https://user-images.githubusercontent.com/29442029/192184189-dd4f230f-474a-4200-9ec9-ee7aeb1b37e4.png)
> > After researching a lot about the approach of workers in our use case, I ended up finding [this](https://stackoverflow.com/a/48484593) answer from StackOverflow which seemed to me to be the most detailed in relation to the use of semaphore vs workers, in it, it is said:
> > ```
> > ..the idiomatic way of limiting the number of parallel tasks this with asyncio is using asyncio.Semaphore
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But this also helped me to understand that `sleep` was not necessary, just using the semaphore as a context manager is enough (removing `sleep` also brought a considerable performance gain).
> 
> I don't like idea, that python have to create several thousands of asyncio objects in memory (and also handle them in parallel). From your link, `...in some cases this approach can have advantages over a semaphore, specifically when the work to be done is very large...`
> 
> I would like to limit concurency at creation of tasks.

Got it. 
I replaced `asyncio.Semaphore` with a solution where we limit the creation of objects and the number of tasks (using `return_when=asyncio.FIRST_COMPLETED`). This seems to me a very clean solution that does exactly what it needs to do.